soroban-env-host/src/host/metered_vector.rs

```diff
@@ -1,250 +1,298 @@
-use super::{MeteredClone, MeteredCmp};
-use crate::xdr::ScHostObjErrorCode;
+use soroban_env_common::{xdr::ScHostFnErrorCode, Compare};
+
+use super::MeteredClone;
 use crate::{
-    budget::{Budget, CostType},
-    HostError,
+    budget::{AsBudget, Budget, CostType},
+    xdr::ScHostObjErrorCode,
+    Host, HostError,
 };
-use im_rc::{vector::Iter, vector::IterMut, Vector};
-use std::cmp::{min, Ordering};
-use std::ops::RangeBounds;
-use std::rc::Rc;
-
-pub struct MeteredVector<A> {
-    budget: Budget,
-    vec: Vector<A>,
+use std::{cmp::Ordering, ops::Range};
+
+#[derive(Clone)]
+pub struct MeteredVector<A>
+where
+    A: MeteredClone,
+{
+    vec: Vec<A>,
 }
 
-impl<A> MeteredVector<A> {
-    fn charge_new(&self) -> Result<(), HostError> {
-        self.budget.charge(CostType::ImVecNew, 1)
+impl<A> MeteredVector<A>
+where
+    A: MeteredClone,
+{
+    fn charge_new(&self, size: usize, budget: &Budget) -> Result<(), HostError> {
+        budget.charge(CostType::VecNew, size as u64)
     }
 
-    fn charge_mut_access(&self, x: u64) -> Result<(), HostError> {
-        self.budget.charge(CostType::ImVecMutEntry, x)
+    fn charge_access(&self, count: usize, budget: &Budget) -> Result<(), HostError> {
+        budget.charge(CostType::VecEntry, count as u64)
     }
 
-    fn charge_immut_access(&self, x: u64) -> Result<(), HostError> {
-        self.budget.charge(CostType::ImVecImmutEntry, x)
+    fn charge_scan(&self, budget: &Budget) -> Result<(), HostError> {
+        budget.charge(CostType::VecEntry, self.len() as u64)
     }
 
-    fn charge_cmp(&self, x: u64) -> Result<(), HostError> {
-        self.budget.charge(CostType::ImVecCmp, x)
+    fn charge_binsearch(&self, budget: &Budget) -> Result<(), HostError> {
+        let mag = 64 - (self.len() as u64).leading_zeros();
+        budget.charge(CostType::VecEntry, mag as u64)
```

Comment:
Nit: I think if we want to ensure over-estimate, we should charge `mag+1`.

Reply:
Oh, sure. Fixed.